modules/effects/arrow-effects-kotlinx-coroutines-instances/src/test/kotlin/arrow/effects/DeferredTest.kt

```diff
@@ -170,35 +170,67 @@ class DeferredKTest : UnitSpec() {
-      forAll(Gen.string(), Gen.list(Gen.string())) { x, xs ->
+      forAll(Gen.string()) { x ->
         runBlocking {
-          checkAwaitAll(ListK.functor(), ListK.traverse(), xs.k()) &&
-            checkAwaitAll(NonEmptyList.functor(), NonEmptyList.traverse(), NonEmptyList(x, xs)) &&
-            checkAwaitAll(Option.functor(), Option.traverse(), Option.just(x)) &&
+          checkAwaitAll(Option.functor(), Option.traverse(), Option.just(x)) &&
             checkAwaitAll(Try.functor(), Try.traverse(), Try.just(x))
         }
       }
     }
 
+    "DeferredK bracket cancellation should release resource with cancel exit status" {
+      runBlocking {
+        lateinit var ec: ExitCase<Throwable>
+        val countDownLatch = CountDownLatch(1)
+        DeferredK.just(Unit)
+          .bracketCase(
+            use = { DeferredK.async<Nothing> { _, _ -> } },
+            release = { _, exitCase ->
+              DeferredK {
+                ec = exitCase
+                countDownLatch.countDown()
+              }
+            }
+          )
+          .value().run {
+            async(Dispatchers.Default) {
+              delay(10)
+              cancel()
+            }
+            withContext(Dispatchers.Default) {
+              k().unsafeAttemptSync()
+            }
+          }
+
+        countDownLatch.await(50, TimeUnit.MILLISECONDS)
+        ec shouldBe ExitCase.Cancelled
+      }
+    }
+
     "DeferredK should cancel KindConnection on dispose" {
       runBlocking {
-        Promise.uncancelable<ForDeferredK, Unit>(DeferredK.async()).flatMap { latch ->
-          DeferredK {
-            DeferredK.async<Unit>(start = CoroutineStart.DEFAULT) { conn, _ ->
-              conn.push(latch.complete(Unit))
-            }.cancel()
-          }.flatMap { latch.get }
-        }.await()
+        val promise = Promise.unsafeUncancelable<ForDeferredK, Unit>(DeferredK.async())
+        val d = DeferredK.async<Unit>(ctx = Unconfined) { conn, _ ->
+          conn.push(promise.complete(Unit))
+        }.unsafeRunAsyncCancellable { }
+          .invoke()
+        promise.get.await()
       }
     }
 
     "KindConnection can cancel upstream" {
-      Try {
-        DeferredK.async<Unit> { conn, _ ->
-          conn.cancel().unsafeRunAsync { }
-        }.unsafeRunSync()
-      }.fold({ e -> e should { it is arrow.effects.ConnectionCancellationException } },
-        { throw AssertionError("Expected exception of type arrow.effects.ConnectionCancellationException but caught no exception") })
+      Promise.uncancelable<ForDeferredK, Unit>(DeferredK.async()).flatMap { latch ->
+        DeferredK.async<Unit> { conn, cb ->
+          conn.push(latch.complete(Unit))
+          cb(Right(Unit))
+        }.flatMap {
+          DeferredK.async<Unit> { conn, _ ->
+            conn.cancel().fix().unsafeRunAsync { }
+          }
+        }.unsafeRunAsyncCancellable { }
+
+        latch.get
```

Comment:
Can we add a timeout here to not block the tests in case of failure?

Reply:
The problem we have now is that we're kind of in an in-between situation where we already have a lot of `cancellable` logic but we don't have any typeclass support for it yet. So all the current concurrency primitives like `Ref`, `Promise`, `Semaphore` etc all have non-cancellable operations since that's the only support we currently get by typeclasses. This is why I am trying to merge these PRs asap so I can write proper laws for this logic, at that point, we will remove all these tests from the concrete types.